File: leetcode/problems/cpp/univalued-binary-tree.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

struct TreeNode {
   int val;
   TreeNode* left;
   TreeNode* right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
public:
   bool isUnivalTree(TreeNode* root) {
      if (!root) return false;
      return check(root, root->val);
   }
   bool check(TreeNode* root, int val) {
      if (!root) return true;
      if (root->val != val) return false;
      return check(root->left, val) && check(root->right, val);
   }
};

File: leetcode/problems/ts/univalued-binary-tree.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// Check if the root and it's left and right child have same value for each node
function isUnivalTree(root: TreeNode | null): boolean {
  if (!root) return true;
  const leftValue = root.left?.val ?? root.val;
  const rightValue = root.right?.val ?? root.val;
  if (root.val !== leftValue || root.val !== rightValue) return false;
  return isUnivalTree(root.left) && isUnivalTree(root.right);
}
